### PR TITLE
Prefetching keyframes and store in memory to improve UX

### PR DESCRIPTION
**Pull request by phongcao**

### CODE REVIEW

File: src/ampPlayer.js [hunk 1]
```diff
@@ -60,9 +60,18 @@ AmpPlayer.prototype.play = function() {
         src: url,
         type: "application/vnd.ms-sstr+xml"
     }]);
+
+    // Prefetch thumbnails on click
+    amPlayer.addEventListener("click", () => {
```
Comment (juzuluag) [inline src/ampPlayer.js]: Can you add an anonymous function here, otherwise Christian's work would break.

File: src/prefetch.js [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    });
+}
+
+function prefetchThumbnails(baseUrl, bandwidth, segments, initializationSegment, initializationSegmentUrl, completeMimeType, memoryCache) {
+    const fetchInterval = 100;
+
+    console.log("Number of segments: " + segments.length);
+    for (var i = 0; i < segments.length; i++) {
+        const time = segments[i].start;
+        setTimeout(() => {
```
Comment (juzuluag) [inline src/prefetch.js]: Same comment with regards anonymous functions.

File: src/prefetch.js [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+                updateSourceBuffers(mediaSource, sourceBuffer, segmentData, offscreenVideoElement);
+            }
+        }, false);
+
+        // Attach the MSE object to the video element
+        offscreenVideoElement.src = URL.createObjectURL(mediaSource, { oneTimeOnly: true });
+        offscreenVideoElement.play().then(_ => {
+            canvasContext.drawImage(offscreenVideoElement, 0, 0, CANVAS_WIDTH, CANVAS_HEIGHT);
+            resolve(canvas.toDataURL());
+        }).catch(error => {
```
Comment (juzuluag) [inline src/prefetch.js]: Anonymous func

**Review by willzhan — APPROVED** [review]

**Review by juzuluag — COMMENTED** [review]
Some comments with regards anonymous func so the build continues working with min changes

File: src/mseutils.js [hunk 4]
```diff
@@ -8,48 +8,59 @@
...
+        // Register sourceopen event handler in order to add source buffers to MSE after it has been attached to the video element.
+        mediaSource.addEventListener("sourceopen", function () {
+            // Register timeupdate event handler to monitor buffer level
+            videoElement.addEventListener("ended", saveThumbnail, false);
+            if (mediaSource.sourceBuffers.length === 0) {
+                // Add video source buffers
+                vidSourceBuffer = createSourceBuffer(mediaSource, completeMimeType, index, videoElement);
+
+                // add segments
+                appendInitSegment(vidSourceBuffer, initializationSegment, initializationSegmentUrl);  //moved here from line 648
```
Comment (juzuluag) [inline src/mseutils.js:35]: Maybe this comment is not longer valid?

File: src/static/js/ampPlayer.js [hunk 5]
```diff
@@ -60,9 +60,18 @@ AmpPlayer.prototype.play = function() {
         src: url,
         type: "application/vnd.ms-sstr+xml"
     }]);
+
+    // Prefetch thumbnails on click
+    amPlayer.addEventListener("click", () => {
```
Comment (juzuluag) [inline src/static/js/ampPlayer.js]: Another anonymous func.

**Review by juzuluag — COMMENTED** [review]
Just more small comments. The rest looks good to me. Good work!

**Review by willzhan — DISMISSED** [review]

File: src/ampPlayer.js (hunk 1 above)
Comment (roblesch) [reply src/ampPlayer.js]: +1, current build script uses `[uglify-js](https://www.npmjs.com/package/uglify-js)` which only supports ES5. To support ES6 code we would need to transpile with something like babel

File: src/mseutils.js [hunk 6]
```diff
@@ -8,48 +8,59 @@
...
+        // Attach the MSE object to the video element
+        videoElement.src = URL.createObjectURL(mediaSource, { oneTimeOnly: true });
+        if (previewType === "video") {
+            var playPromise = videoElement.play();
+            if (playPromise !== undefined) {
+                playPromise.then(_ => {
+                    // Automatic playback started!
+                    // Show playing UI.
+                    // We can now safely pause video...
+                    //video.pause();
```
Comment (roblesch) [inline src/mseutils.js]: Nit: this looks like it can be removed

File: src/mseutils.js [hunk 7]
```diff
@@ -8,48 +8,59 @@
...
+                })
+                    .catch(error => {
+                        // Auto-play was prevented
+                        console.log("Promise for video.play() is rejected.");
+                    });
+            }
```
Comment (roblesch) [inline src/mseutils.js]: Nit: whitespace

File: src/prefetch.js [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+ * This module provides functionality to prefetch keyframes and store them in memory.
+ */
+
+// Canvas
+const CANVAS_WIDTH = 200;                               // canvas width
+const CANVAS_HEIGHT = 90;                               // canvas height
+
+let canvas;                                             // canvas used for offscreen rendering
+let canvasContext;                                      // canvas context
+let offscreenVideoElement;                              // used to decode video segment
```
Comment (roblesch) [inline src/prefetch.js]: Use ES5 `var`

File: src/prefetch.js [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+            } else {
+                reject("Failed to download segment!");
+            }
+        });
+    });
+}
+
+// Convert video segment to base64 PNG image
+function segmentToImage(initializationSegment, initializationSegmentUrl, completeMimeType, segmentData) {
+    return new Promise((resolve, reject) => {
```
Comment (roblesch) [inline src/prefetch.js]: ES5 does not support `Promise`. I will look into using babel for transpiling

File: src/mseutils.js (hunk 4 above)
Comment (phongcao) [reply src/mseutils.js:35]: I think that it's still valid. We do add both init segment and the downloaded one.

**Review by juzuluag — DISMISSED** [review]
LGTM

**Review by roblesch — DISMISSED** [review]

**Review by juzuluag — DISMISSED** [review]

**Review by juzuluag — APPROVED** [review]
nice!

**Review by dnastrain — APPROVED** [review]

**Pull request merged**